Frontend Translation + bug fixes + tweakings

LordWilbur opened the pull request:
- fixed an error on finance in admin panel
- flight ident now use this format: VA IATA(if empty ICAO) + Flight Number + - + Flight Code (without C) + - + Flight Leg (without L)
- added function __trans_choice in helpers.php for translation
- fixed error in flight edit/insert panel not showing/inserting Tuesday in days
- fixed an error occurring when metar retrieved is empty
- edited now-ui-kit.css to align login fields correctly
- added /public/assets/frontend/js/core/jquery-3.3.1.min.js to fix a missed resource error in authentication pages
- added translations file for en and it locales
- translated all the frontend templates

nabeelio commented on app/helpers.php:
```diff
@@ -221,3 +221,19 @@ function _fmt($line, array $replace)
+     * @param null $locale
+     * @return string
+     */
+    function __trans_choice($key, $number, array $replace = [], $locale = null)
```
Can we change this to `__tc`, just to make it shorter in the templates?

LordWilbur replied:
if i move all in php files then there is a dedicated function called trans_choice()

nabeelio replied:
Ok. Don't change anything just yet in your PR, let me do some research

nabeelio commented on public/assets/frontend/css/now-ui-kit.css:
```diff
@@ -2089,7 +2089,6 @@ fieldset[disabled]
 
 .form-group .form-control,
 .input-group .form-control {
-  margin-top: 2px;
```
This needs to be changed in the SASS file and then rebuilt - it's in `resources/sass/now-ui` , and then `make build-assets`

nabeelio commented:
Hey, this looks awesome! A few notes/questions:

1. The build failed:

```
RuntimeException: Translation file [/home/travis/build/nabeelio/phpvms/resources/lang/en.json] contains an invalid JSON structure.
```
To run the tests locally, use

```sh
make tests
```

2. I think it's better to have short translation strings - so instead of `"Fill out the captcha": "Fill out the captcha", making it: `"captcha.fill": "Fill out the captcha",`

Which is how I've done it in the past - what do you think? It keeps the templates smaller and in my experience, if the translation strings look just like what's being put out, someone will edit it and then wonder why it may be broken. Though I think Laravel just outputs that same string? What're your thoughts? Is this the way it's normally done with Laravel?

3. Using JSON vs PHP - right now everything is in the `.php` files, instead of using JSON, I think we should keep it as PHP. I like consistency. Being able to add comments might be good as well.

4. There's an airplane image included - was that by accident?

nabeelio commented:
In terms of database content, we just won't be able to translate. Since it's usually data that the user has entered in, it'll just be in whatever language they use. There are some things like the settings that are in the install files, though. Maybe those descriptions can be translated.

nabeelio commented on app/Http/Controllers/Admin/FinanceController.php:
```diff
@@ -51,7 +51,7 @@ public function index(Request $request)
         foreach ($airlines as $airline) {
             # Return all the transactions, grouped by the transaction group
-            $transactions = JournalTransaction::groupBy('transaction_group')
+            $transactions = JournalTransaction::groupBy('transaction_group', 'currency')
```
I think this `currency` will break things - what was the exact issue? Do the tests still pass?

LordWilbur replied:
the page showed a sql exception saying that curerncy field was not in group by clause

nabeelio replied:
Can you tell me how to duplicate this? It might be a bug in the package I got this from as well, I'll have to look

LordWilbur replied:
it happened everytime i clicked on finance in admin panel. Anyway you can't select a field in a query where there are aggregate functions without putting that field in the group by clause too, it's a sql rule....

nabeelio replied:
Hmm, I'm not running into it. Can you export your database where it's happening and send it to me? I just want to understand exactly what's going on. Grouping by the `transaction_group` should be enough to limit it. What version of MySQL are you on?

nabeelio replied:
Yes, I know that, I've been doing SQL for 10+ years lol. There's already a group and the query should be correct. Adding a second group could make the results actually be incorrect, and the first query you outlined there is what's actually being created. 

So, I'm trying to understand why I'm not seeing an error, but you are. Because if I'm not seeing it but you are, then there's something else that is wrong that's going on, which is what I want to understand, before changing anything. There are a *lot* of small incompatibilities between Maria and MySQL.

nabeelio replied:
If you can send me a dump of your database so I can load it and see if I get the error, that would be appreciated. And then I'm doing to install Maria and then Percona to see if I get it on those. This needs to work for everyone, so it's a little more involving on the testing side.

LordWilbur replied:
What i’m trying to say is that all non aggregate fields must be in the group by clause if there is some aggregate function in the query (and this is a basic and strict rule for all dbs i’ve worked with in the 16 years i’ve made as software developer). 

The fact that MySQL allows that kind of queries is an error on 99% of the other relational dbs around.

Now i’m at work and don’t have my laptop bere so i can’t send tuo the fumo of my db, but i’ll do that tonight when i’m back home.

Anyway you can execute both queries in MySQL and you can see that results are identical.

I find your project very well coded and organized, but, if i can say, the db side of the app is a little bit confusing and difficult to understand and browse (as said before, i’ve worked 16 years as developer, but i’ve many difficulties to understand and find stuff in your db code).

Tonight, or latest tomorrow morning i’ll submit my new work again (you’d only make the assets cause i haven’t found a solution for that on windows). If all is ok than i go straight forward to complete the translation of the app and after that i can look if it’s possibile to reorganize and optimize the db infrastructure, if you want...

nabeelio replied:
Yes, I understand how it's supposed to work and I know how SQL works. My point is that we're seeing *different behavior for the same code*. The `groupBy('transaction_group')` is supposed to be generating the correct query, and if it's not then there's a problem either in the ORM layer, or there is difference between Maria and MySQL. I'm trying to understand why exactly the original code isn't working *before* changing anything. Without understanding why it's broken or doesn't work can cause issues later. Remember, this needs to work for *everyone*. 

I have never had to put any non-aggregate fields into the `GROUP BY` and that would cause different behavior. For example, in this case, the currency is set to `USD`, with a transaction group of `Gate Costs`, for example. If you add a second group-by, if there are transactions of a different currency, they will be grouped separately, which is not what we want in this case.

If you do an `EXPLAIN`, the underlying query will be different. I haven't tried it yet. But I want to understand why we're seeing different behavior. Does that make sense?

We can talk on Discord about the DB stuff. It's using a standard Repository-pattern on top of the ORM/models layer. Most queries go through the repository layer first, in order to take advantage of caching, without having to wrap individual calls to the models in caching. The repository also enables some rudimentary searching and filtering, and also allows caching these.

nabeelio replied:
And what's your value for this:

```sql
SELECT @@SQL_MODE, @@GLOBAL.SQL_MODE;
```

nabeelio replied:
I'm just trying to see why exactly MySQL is treating this differently and it's version specific. It's basically for my own edification. 

If you can fix the PR so the tests are also passing, I'll merge it all in.

LordWilbur replied:
@@SQL_MODE = STRICT_TRANS_TABLES,ERROR_FOR_DIVISION_BY_ZERO,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION

@@GLOBAL.SQL_MODE = STRICT_TRANS_TABLES,ERROR_FOR_DIVISION_BY_ZERO,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION

Here is my dump:

[Dump12052018.zip](https://github.com/nabeelio/phpvms/files/2014820/Dump12052018.zip)

So seems to me that the error is somewhere in the app's db layer or in the MySQL PDO cause on mariaDB it generates this query:

SELECT transaction_group, currency, [...] GROUP BY transaction_group

and it's obviously wrong....

nabeelio replied:
Thanks. Yeah I want to just understand if it's a problem in the ORM or in PDO. Because if there's a bug here then it'll be doing the same in other calls. Thanks!

nabeelio commented:
Regards to the jquery file - update the auth template to use the one that's already there. Let's not add another jquery version/file, it gets confusing. Also, those assets are pulled in via `npm` in the `package.json` file, and then copied over using the webpack file.

LordWilbur commented:
i cannot run make cause i'm on windows....
i'll move everything in php files then, but in this way i find it was easier for users to understand which string they were translating.
For the __trans_choice function i must see how laravel works with php files but renaming it is no problem for me.

nabeelio commented:
Can you run artisan? The tests are just using PHPUnit, I'm sure they can be run on Windows. If you make the changes to the SASS file, when we merge it, I'll rebuild the assets. Or instead of `__trans_choice`, call it `__trans`, since that makes it more clear that it's a translation.

And ok, let's leave the full string for translations, the way you have it now.

LordWilbur commented:
regard jquery the problem was that in the template there was jquery.3.2.1-min.js but there was no file in the assets...

nabeelio commented:
Yeah, I don't mind it being in JSON files, put with some things in JSON and some things in PHP, I think they should all be one thing or another.

nabeelio commented:
Yeah, I understand that. The template should be updated to use the jquery that's currently there in `assets/global/js/jquery.js`

LordWilbur commented:
sorry didn't understand what you meant, anyway that is simple enough ;)

nabeelio commented:
What's the refactoring you're doing? Just moving them into the PHP files? You can leave the translations as-is, I think the way you did that is fine (using the full sentences, etc. I misunderstood something on the Laravel level)

nabeelio commented on app/Models/Flight.php:
```diff
@@ -112,12 +112,12 @@ public function getIdentAttribute(): string
-        if (filled($this->route_code)) {
-            $flight_id .= '/C'.$this->route_code;
+        if (filled($this->route_leg)) {
+            $flight_id .= '-'.$this->route_leg;
```
I'm a little torn on this change. It's hard to tell what the code/leg are, especially if you need to enter into an ACARS app. I'll leave it for now but might change this later.

LordWilbur replied:
normally code identifies the type of transport that the airplane makes ( for example X is Economy Class)
route leg is a numeric value that identifies the step in a long flight with multiple stops (for example a flight from San Francisco to Rome should stop in New York, Barcellona and Paris. So it mantains the same flight number but San Francisco-New York is leg 1, New York-Barcellona is leg 2, Barcellona-Paris is leg 3 and finally Paris-Rome is leg 4)

nabeelio replied:
The leg is correct, though I can't think of a time that I've seen the leg be used (at least with the airlines I've worked with. I write operations software for GE Aviation). RE the code - what you're talking about is the fare class; that's shown separately. The way those are coded at `[CLASS][CAPACITY or TOTAL]`, so "F10B20Y100" would be for that one. I have that noted in an issue to add that in.  For the flight itself, the code is actually  the revenue coding, which is only used internally for codeshares, etc. So it's not really important in terms of identifying flights, only the flight number and leg (though this is simplified too from actually operations)

LordWilbur commented:
On my db now i have 2 airlines registered with ICAO PVA and SIC.
I've registered 3 users 2 on PVA and 1 on SIC and i was expecting that the system gave them PVA0001, PVA0002 and SIC0001 but i have PVA0001, SIC0002 and PVA003. So i have seen that in the users table tha callsign of each pilot isn't saved but generated from the user_id.
My questions are: 

1) Isn't it better if we generate the callsign (only if pilot auto accept is true) adding 1 to the highest already registered and save it in the db allowing admins to change that too eventually? 

2) is possible to change the login behaviour to make users login with their callsign instead of their email (to me seems more "professional")?

nabeelio commented:
What IDE do you use? IntelliJ will be able to run the tests for you.

I have both of those changes done in a local branch, actually. I'll clean it up and merge it in soon.

LordWilbur commented:
Very nice... So my ideas weren't so bad ;)

nabeelio commented:
lol, not at all! I have a bunch of branches locally that I need to rebase and fix up. That one is probably a month old at this point.

LordWilbur commented:
Ok finally i have managed to rebuild assets and pass all travis tests ;)

nabeelio commented:
Looks good! It just needs to be updated against the latest dev, so once you do that I will merge it in. And just the one comment

nabeelio commented on app/Database/migrations/2018_05_16_200138_change_route_leg_to_unsigned_integer.php:
```diff
@@ -0,0 +1,32 @@
+    public function up()
+    {
+        Schema::table('flights', function (Blueprint $table) {
+            $table->unsignedInteger('route_leg')->nullable()->change();
```
I'll move this into the main migration. I'm requiring re-installs for right now since there would be too many migrations

nabeelio commented:
You have to do a `git pull --rebase origin dev` so it'll merge in properly. It looks like other updates were copied in?

Also, just change all tabs to spaces. Trying to stay with the PSR formatting and avoiding the situation in phpVMS legacy where half the files are tabbed and the other half are spaced :)

nabeelio commented on resources/lang/en/frontend.php:
```diff
@@ -0,0 +1,266 @@
+
+return [
+	'global' =>[
+		'dashboard' => 'Dashboard',
```
Oh... I said you should have left the original strings in but just move it to PHP. We can revert the commit so you don't have to manually undo anything and I'll move the strings into the PHP files. I feel bad that you've done all this work! Lol

nabeelio commented:
Ok here's what I think I'll do. I won't merge it directly, I'll put the branch down to my machine and cherry-pick all of the commits to bring in. I think trying to fix the rebase will be a nightmare. I'll work on doing that right now.

nabeelio commented:
I'm going to have to also manually copy some changes in. Just so you know for the future, always separate commits for features. So the changes for the route leg should all be in one commit, all the translation stuff in another commit.

Since I'm manually moving changes, I'll put your name in the commits so you still get the credit for them

nabeelio commented on resources/lang/en/frontend.php:
```diff
@@ -87,6 +87,33 @@
 		'nodownloads' => 'There are no downloads!',
 	],
 	
+	'emails' => [
```
Wait I said to leave these as the full strings as you were doing in the JSON file, just put them in PHP

nabeelio commented:
It's much more difficult because all your commits are combined, I can't even go through them in order because there are too many conflicts.

Also I said to leave the translations as the original strings... it's fine like this, though. I was looking at PhiloNL/laravel-translate so this might still work with it.

nabeelio commented:
There are too many conflicts to merge. If you can, check out a fresh copy of dev, and copy *only* the templates and language changes and create a new pull request. I'm getting conflicts in every file just trying to merge between two different commits.

LordWilbur commented:
I've rebased all the branch on dev, removed all tabs from files, deleted my migration file and moved the change in that for the flight table. 

now it should be easier for you to commit what you need

nabeelio commented:
Ok. I'm going to close this and clean it up

nabeelio commented:
Btw please don't make any other changes until I get these in and formatted. Should be done sometime tomorrow or Sunday.

nabeelio commented:
Hey, it's all merged in. I split up the base translations into different files, so I removed the frontend file, and I also removed the system file I made. I sorted some of the common words and phrases into that common file as well.
